harness/determined/common/api/authentication.py

```diff
@@ -62,139 +69,79 @@ def default_load_user_password(
-            token is None
-            and util.get_det_username_from_env() is not None
-            and util.get_det_user_token_from_env() is not None
-            and requested_user in (None, util.get_det_username_from_env())
-        ):
-            session_user = util.get_det_username_from_env()
-            assert session_user
-            token = util.get_det_user_token_from_env()
-
-        if token is not None:
-            return UsernameTokenPair(session_user, token)
-
-        # We'll need to create a new token, so we'll need a password. If there was no requested
-        # user and we ended up falling back to the default username `determined`, then we can fall
-        # back to the default login as well. Otherwise, ask the user for their password.
-        fallback_to_default = password is None and session_user == constants.DEFAULT_DETERMINED_USER
-        if fallback_to_default:
-            password = constants.DEFAULT_DETERMINED_PASSWORD
-
-        if password is None:
-            password = getpass.getpass("Password for user '{}': ".format(session_user))
-
-        try:
-            token = do_login(self.master_address, session_user, password, cert)
-        except api.errors.ForbiddenException:
-            if fallback_to_default:
-                raise api.errors.UnauthenticatedException(username=session_user)
-            raise
-
-        self.token_store.set_token(session_user, token)
-
-        return UsernameTokenPair(session_user, token)
-
-    def is_user_active(self, username: str) -> bool:
-        return self.token_store.get_active_user() == username
-
-    def get_session_user(self) -> str:
-        """
-        Returns the session user for the current session. If there is no active
-        session, then an UnauthenticatedException will be raised.
-        """
-        return self.session.username
-
-    def get_session_token(self, must: bool = True) -> str:
-        """
-        Returns the authentication token for the session user. If there is no
-        active session, then an UnauthenticatedException will be raised.
-        """
-        if self.session is None:
-            if must:
-                raise api.errors.UnauthenticatedException(username="")
-            else:
-                return ""
-        return self.session.token
-
-
-def do_login(
-    master_address: str,
-    username: str,
-    password: str,
+def login_with_cache(
+    master_address: Optional[str] = None,
+    requested_user: Optional[str] = None,
+    password: Optional[str] = None,
     cert: Optional[certs.Cert] = None,
-) -> str:
-    password = api.salt_and_hash(password)
-    unauth_session = api.Session(user=username, master=master_address, auth=None, cert=cert)
-    login = bindings.v1LoginRequest(username=username, password=password, isHashed=True)
-    r = bindings.post_Login(session=unauth_session, body=login)
-    token = r.token
-
-    return token
-
-
-class LogoutAuthentication(Authentication):
+) -> UsernameTokenPair:
```

Comment:
why do we need the UsernameTokenPair object at all? i suspect there is a good reason, i'm just curious.

i would have thought these login methods would return a Session

Reply:
Hm, I did think about giving the UnauthSession a login() method that returns a Session.

I guess the reason I didn't investigate that possibility very hard was mostly practical.

In a refactor, I think it's hard enough to untangle bad abstractions, but it's even harder to then combine things into the ideal abstractions all in one pass.  I've tried plenty of times, and usually I end up three quarters through migrating to my new and perfect abstraction, when I realize there's some dusty corner of the codebase that I don't want to rewrite that depends on some detail I thought I could abstract away.

So for large refactors, I think it's much easier to untangle things "too much" into too small of abstractions in the first pass.  Migrate the whole codebase to that, and after everything is up and running again, you can make a second pass (usually a quick one) that combines the smaller abstractions in a way that makes more sense.  Basically, entangling things is always way faster than untangling them.

Reply:
I lost my previous response in a browser mishap, but I'm in agreement with Anda. (Also think it could be a follow-up commit if you like that better.)

Logging in only makes sense in the context of a master, so if we're going to return an object that contains context, I think a Session is the best choice.

Also possible is a token, but the logic here with inferring users is complicated, and I get why doing that in a way that makes sense seems too hard.

Reply:
I keep seeing the pattern

```python
       cert = certs.default_load(info.master_url)
       utp = authentication.login_with_cache(info.master_url, cert=cert)
       sess = api.Session(info.master_url, utp, cert)
```

This feels like good evidence to me that a `Session` is the reason to log in, and that `utp` is an intermediate product. If a caller really wants a `username` or `token`, those are right there in the session.

I also think the necessary replication of `cert` and `master_url` is evidence that `utp` isn't quite the right abstraction.

Reply:
This is a good point, but I don't plan on doing it as part of this PR, specifically so I can land this before I go on parental leave.

Reply:
Added to the wiki.